### PR TITLE
Changed `Extensions` and `Mimes` in `FormatDescriptor` from `string[]` to `ReadOnlyCollection<string>`

### PR DESCRIPTION
**Pull request by Lehonti**
Also, I added an extension method that turns `IEnumerable<T>` into `ReadOnlyCollection<T>`, for convenience and readability.

This is intended to make it clear from the public interface of the class whether or not the members of the collection should be modified. I intend to do the same in more classes, in future pull requests.

### CODE REVIEW

File: Pinta.Core/ImageFormats/FormatDescriptor.cs [hunk 1]
```diff
@@ -73,22 +75,24 @@ public sealed class FormatDescriptor
...
 					 IImageImporter? importer, IImageExporter? exporter)
 		{
-			if (extensions == null || (importer == null && exporter == null)) {
+			if (extensions == null || (importer == null && exporter == null))
 				throw new ArgumentNullException ("Format descriptor is initialized incorrectly");
-			}
 
-			this.Extensions = extensions;
-			this.Mimes = mimes;
+			var materializedExtensions = extensions.ToReadOnlyCollection ();
```
Comment (cameronwhite) [inline Pinta.Core/ImageFormats/FormatDescriptor.cs:84]: Are these separate local variables needed, or can the subsequent code just use `this.Extensions` instead?

File: Pinta.Core/Extensions/OtherExtensions.cs [hunk 2]
```diff
@@ -9,11 +9,38 @@
...
+		{
+			switch (values) {
+				case ImmutableBackedReadOnlyCollection<T> transparent:
+					return transparent;
+				case ImmutableArray<T> array:
+					return new ImmutableBackedReadOnlyCollection<T> (array);
+				case ImmutableList<T> list:
+					return new ImmutableBackedReadOnlyCollection<T> (list);
+				default:
+					return new ImmutableBackedReadOnlyCollection<T> (values.ToImmutableArray ());
```
Comment (cameronwhite) [inline Pinta.Core/Extensions/OtherExtensions.cs:30]: In the `FormatDescriptor` use case, there aren't any methods which will modify the lists so I feel like just having a plain `ReadOnlyCollection` wrapping the list is probably fine?
I'm not sure I see the need for doing an extra copy into an immutable array

**Comment by Lehonti** [comment]
@cameronwhite I understand what you are saying. Allow me to explain my reasoning so that you can make a decision.

tl;dr: I'm trading compiler-enforced immutability for a few extra allocations.

In the original code, the references to the array arguments were being assigned to the properties directly. This would allow someone to modify what is supposed to be the state of the object by manipulating the arrays, which I don't think is a good practice. Thus, a snapshot of the collection is made. This of course introduces a new memory allocation, but it is not at all uncommon, especially in library functions. I know that, in practice, the constructor is only being called with arrays created at the point of invocation, but this makes the intent clearer.

Since I've determined (in the previous step) that a copy of the array elements will be made one way or the other, and I see that it will never be modified, I may as well create an `ImmutableArray<T>` instead of a regular array `T[]`; but notice that I'm doing it behind the scenes, in an extension method, using a private collection type that inherits from `ReadOnlyCollection<T>` but that nobody will ever see from the outside. This gives information about its immutability at the type level. The extra allocations only happen the first time that this extension method is called. If they were to call the method a second time, the pattern matching would detect it and just return the same reference, with no extra allocations being made. I'm defensively writing this extension method so that it can be used in other parts of the project that expose collection properties, not just `FormatDescriptor`.

Finally, to understand why the separate variables were added, notice that I changed the signature of the constructor. Instead of getting `string[]` they are now getting `IEnumerable<string>`. At the same time, I am trying to avoid:
- Iterating the _source_ `IEnumerable`s twice. They are being iterated in `foreach` loops further down in the body of the constructor. I want to ensure that they will only be iterated once, as iteration may be expensive or someone may pass a malicious `IEnumerable` that gives a different sequence of items the second time.
- Relying on the current state of the properties. Before they are assigned a value, `Extensions` and `Mimes` are null.

My reasoning may sound too strict, but it comes from experience. I've used these principles to refactor code that seemed untractable.

**Comment by cameronwhite** [comment]
Thanks!
- I think that's a fair argument for the immutable array and I'm okay with that change - perhaps you can add some of the info from your comment into the code comments so it's recorded there?
- For the second part, maybe there'd be a compiler warning if `this.Extensions` was read before being initialized? Since it's not a super complicated method I feel it's probably the most readable to do something like
```csharp
this.Extensions = extensions.ToReadOnlyCollection();
...
foreach (string ext in this.Extensions)
...
```
versus adding a third variable that has some version of the list :)

File: Pinta.Core/ImageFormats/FormatDescriptor.cs [hunk 3]
```diff
@@ -78,21 +78,21 @@ public sealed class FormatDescriptor
 		public FormatDescriptor (string displayPrefix, IEnumerable<string> extensions, IEnumerable<string> mimes,
 					 IImageImporter? importer, IImageExporter? exporter)
 		{
-			if (extensions == null || (importer == null && exporter == null))
-				throw new ArgumentNullException ("Format descriptor is initialized incorrectly");
+			if (extensions is null)
```
Comment (cameronwhite) [inline Pinta.Core/ImageFormats/FormatDescriptor.cs:81]: I just noticed this null check is probably unnecessary since we build with nullable types enabled and  `extensions` is not nullable

**Comment by cameronwhite** [comment]
Thanks, all the changes look good
I just noted one additional fix, and the formatting check is also failing (maybe from line endings?)

**Comment by Lehonti** [comment]
I just removed the redundant null check and ran `dotnet format` to see if anything came up. Apparently several files were changed by it.

I don't think that the formatting check is failing from the changes in _this_ pull request. I just checked all the line endings and they are all CRLF. The checks also failed in a previous one (where I just couldn't find the problem so I gave up and closed it).

**Comment by Lehonti** [comment]
Yeah, I don't quite understand why, but it seems that normalizing the line endings in the problematic files to `LF` solved the formatting issue, even though most (all?) of these files weren't originally involved in this pull request. 🤷‍♂️

**Comment by cameronwhite** [comment]
Sorry for the formatting troubles! 😢  The `master` branch CI build was passing this okay, but perhaps something with git's line ending conversion messed things up for you
